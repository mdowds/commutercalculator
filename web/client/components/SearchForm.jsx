import React from 'react';
import PropTypes from 'prop-types';
import Autocomplete from 'react-autocomplete';

export default class SearchForm extends React.Component {

    constructor(props) {
        super(props);
        this.state = {inputText: "", selectedStation: {}};
        this.handleSubmit = this.handleSubmit.bind(this);
    }

    handleSubmit() {
        this.props.onSubmit(this.state.selectedStation);
    }

    suggestStationName(station, inputText) {
        return (
            station.name.toLowerCase().indexOf(inputText.toLowerCase()) !== -1 ||
            station.id.toLowerCase().indexOf(inputText.toLowerCase()) !== -1
        )
    }

    render() {
        const borderStyle = "1px solid grey";

        const acMenuStyle = {
            borderRadius: '3px',
            boxShadow: '0 2px 12px rgba(0, 0, 0, 0.1)',
            background: 'rgba(255, 255, 255, 0.9)',
            fontSize: '90%',
            position: 'fixed',
            overflow: 'hidden',
            maxHeight: 400,
            zIndex: 10,
            border: borderStyle,
        };

        return (
            <div style={{padding: 5, boxSizing: "border-box"}}>
                <span style={{fontSize: "90%", paddingRight: 5}}>Journeys to</span>
                <Autocomplete
                    value={this.state.inputText}
                    items={this.props.destinations ? this.props.destinations : []}
                    getItemValue={station => station.name}
                    renderItem={(station, isHighlighted) => (
                        <div style={{zIndex: 10, padding: 10, borderBottom: borderStyle}} key={station.id}>{station.name}</div>
                    )}
                    menuStyle={acMenuStyle}
                    onChange={(event, value) => this.setState({inputText: value})}
                    onSelect={(value, station) => this.setState({inputText: station.name, selectedStation: station})}
                    shouldItemRender={this.suggestStationName}
                />
                <input id="submitSearch" type="button" value="Go" onClick={this.handleSubmit}/>
            </div>
        );
    }
}

SearchForm.propTypes = {
    onSubmit: PropTypes.func.isRequired,
    destinations: PropTypes.arrayOf(PropTypes.object)
};
