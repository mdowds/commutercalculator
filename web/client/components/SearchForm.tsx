import * as React from 'react';
import * as Autocomplete from 'react-autocomplete';
import styles from './styles/SearchFormStyles';
import SearchFilters from "./SearchFilters";
import {Station, SelectedFilters} from '../types';

interface SearchFormProps {
    onSubmit(selectedStation: Station, filters: SelectedFilters) : void;
    readonly destinations: Array<Station>;
}

interface SearchFormState {
    inputText: string;
    selectedStation?: Station;
    showFilters: boolean;
    filters: SelectedFilters;
}

export default class SearchForm extends React.Component<SearchFormProps, SearchFormState> {

    constructor(props: SearchFormProps) {
        super(props);
        this.state = {inputText: "", selectedStation: undefined, showFilters: false, filters: {}};
        this.handleSubmit = this.handleSubmit.bind(this);
        this.handleFilterChange = this.handleFilterChange.bind(this);
    }

    handleSubmit() {
        if(this.state.selectedStation === undefined) return;
        this.setState({showFilters: false});
        this.props.onSubmit(this.state.selectedStation, this.state.filters);
    }

    handleFilterChange(filters: SelectedFilters) {
        this.setState({filters: Object.assign(this.state.filters, filters)});
    }

    static suggestStationName(station: Station, inputText: string) {
        return (
            station.name.toLowerCase().indexOf(inputText.toLowerCase()) !== -1 ||
            station.id.toLowerCase().indexOf(inputText.toLowerCase()) !== -1
        )
    }

    render() {
        const filters = this.state.showFilters ? <SearchFilters onChange={this.handleFilterChange} /> : null;

        return (
            <div>
                <div style={styles.overallWrapper}>
                    <Autocomplete
                        value={this.state.inputText}
                        items={this.props.destinations ? this.props.destinations : []}
                        getItemValue={station => station.name}
                        renderItem={(station, isHighlighted) => (
                            <div style={styles.acItem} key={station.id}>{station.name}</div>
                        )}
                        inputProps={{placeholder: "Enter destination in Zone 1", style: styles.searchInput}}
                        menuStyle={styles.acMenu}
                        onChange={(event, value) => this.setState({inputText: value})}
                        onSelect={(value, station) => this.setState({
                            inputText: station.name,
                            selectedStation: station
                        })}
                        onMenuVisibilityChange={(isOpen) => this.setState({showFilters: isOpen ? false : this.state.showFilters})}
                        shouldItemRender={SearchForm.suggestStationName}
                        wrapperStyle={styles.acWrapper}
                    />
                    <input id="toggleFilters" type="button" value="Filters" style={styles.filtersButton}
                           onClick={() => {
                               this.setState({showFilters: !this.state.showFilters})
                           }}/>
                    <input id="submitSearch" type="button" value="Go" onClick={this.handleSubmit}
                           style={styles.goButton}/>
                </div>
                {filters}
            </div>
        );
    }
}
