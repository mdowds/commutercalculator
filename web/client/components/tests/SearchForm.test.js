import React from 'react';
import renderer from 'react-test-renderer';
import {shallow} from 'enzyme';
import SearchForm from '../SearchForm.jsx';
import Autocomplete from 'react-autocomplete';

test("SearchForm renders the form with Autocomplete", () => {
    const form = renderer.create(<SearchForm onSubmit={()=>{}} destinations={[]} />);
    expect(form.toJSON()).toMatchSnapshot();
});

test("SearchForm updates inputText when station input changes", () => {
    const form = shallow(<SearchForm onSubmit={()=>{}} destinations={[]} />);
    expect(form.state().inputText).toEqual("");

    form.find(Autocomplete).simulate('change', {}, "foo");
    expect(form.state().inputText).toEqual("foo");
});

test("SearchForm updates selectedStation and inputText when station is selected", () => {
    const dest = {id: "FOO", name: "Foo Station"};
    const form = shallow(<SearchForm onSubmit={()=>{}} destinations={[dest]} />);
    expect(form.state().selectedStation).toEqual({});
    expect(form.state().inputText).toEqual("");

    form.find(Autocomplete).simulate('select', dest.name, dest);
    expect(form.state().inputText).toEqual("Foo Station");
    expect(form.state().selectedStation).toBe(dest);
});

test("SearchForm calls onSubmit when Search button clicked", () => {
    let calledWith = {};
    const dest = {id: "FOO", name: "Foo Station"};

    const form = shallow(<SearchForm onSubmit={x => calledWith = x} destinations={[dest]} />);
    form.setState({selectedStation: dest});
    form.find('#submitSearch').simulate('click');
    expect(calledWith).toBe(dest);
});

test("SearchForm doesn't call onSubmit when Search button clicked with no selected station", () => {
    let submitCalled = false;
    const dest = {id: "FOO", name: "Foo Station"};

    const form = shallow(<SearchForm onSubmit={() => submitCalled = true} destinations={[dest]} />);
    form.find('#submitSearch').simulate('click');
    expect(submitCalled).toEqual(false);
});

test("suggestStationName returns correct results", () => {
    const dest = {id: "ABC", name: "Foo Station"};
    const form = new SearchForm(()=>{} ,[dest]);

    expect(form.suggestStationName(dest, "Foo")).toEqual(true);
    expect(form.suggestStationName(dest, "foo")).toEqual(true);
    expect(form.suggestStationName(dest, "abc")).toEqual(true);
    expect(form.suggestStationName(dest, "ABC")).toEqual(true);
    expect(form.suggestStationName(dest, "f")).toEqual(true);
    expect(form.suggestStationName(dest, "Bar")).toEqual(false);
});
