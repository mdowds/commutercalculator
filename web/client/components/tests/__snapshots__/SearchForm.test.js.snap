// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchForm renders the form with Autocomplete 1`] = `
<div
  style={
    Object {
      "position": "relative",
    }
  }
>
  <div
    style={
      Object {
        "height": 38,
        "width": "100%",
      }
    }
  >
    <input
      aria-autocomplete="list"
      aria-expanded={false}
      autoComplete="off"
      onBlur={[Function]}
      onChange={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      onKeyUp={[Function]}
      placeholder="Enter destination (Zone 1 stations only)"
      role="combobox"
      style={
        Object {
          "border": "1px solid grey",
          "boxSizing": "border-box",
          "fontSize": "90%",
          "padding": 10,
          "paddingRight": 38,
          "width": "100%",
        }
      }
      value=""
    />
  </div>
  <input
    id="submitSearch"
    onClick={[Function]}
    style={
      Object {
        "border": "1px solid grey",
        "height": 38,
        "position": "absolute",
        "right": 0,
        "top": 0,
        "width": 38,
      }
    }
    type="button"
    value="Go"
  />
</div>
`;
